### PR TITLE
Defer bracket pair document parsing until it's actually used

### PR DESCRIPTION
**Pull request by Tyriar**
Part of #161622

This change queues parsing of the document so that it does not block editor input latency. This is done via `setImmediate` such that it is called before any timers, the only downside to this is that changing a character adjacent to a bracket may appear white for a single frame.

This is a good trade off imo to save what I measured sometimes to be > 10% of the keypress event's runtime. The amount of time spent is quite variable, not sure why.

### CODE REVIEW

**Comment by Tyriar** [comment]
@alexdima can I not use set/clearImmediate in monaco?

```
Error: src/vs/editor/common/model/bracketPairsTextModelPart/bracketPairsTree/bracketPairsTree.ts(76,12): error TS233[9](https://github.com/microsoft/vscode/actions/runs/3276846488/jobs/5393380511#step:10:10): Property 'clearImmediate' does not exist on type 'Window & typeof globalThis'.
Error: src/vs/editor/common/model/bracketPairsTextModelPart/bracketPairsTree/bracketPairsTree.ts(123,31): error TS2339: Property 'setImmediate' does not exist on type 'Window & typeof globalThis'.
```

**Comment by alexdima** [comment]
> can I not use set/clearImmediate in monaco?

Unfortunately `setImmediate` is [out of the standards track](https://developer.mozilla.org/en-US/docs/Web/API/Window/setImmediate) and not available on the web. AFAIK it is deprecated and possibly only available in Electron/node.js

**Comment by Tyriar** [comment]
@alexdima good to know, will switch to timeout

**Comment by Tyriar** [comment]
This is where it happens now:

image

Probably better than immediate actually as immediate was happening before the animation frame

File: src/vs/editor/common/model/bracketPairsTextModelPart/bracketPairsTree/bracketPairsTree.ts [hunk 1]
```diff
@@ -105,11 +106,25 @@ export class BracketPairsTree extends Disposable {
...
 
-		this.astWithTokens = this.parseDocumentFromTextBuffer(edits, this.astWithTokens, false);
-		if (this.initialAstWithoutTokens) {
-			this.initialAstWithoutTokens = this.parseDocumentFromTextBuffer(edits, this.initialAstWithoutTokens, false);
+	private queueParsing(edits: TextEditInfo[]): void {
+		// Queues parsing of the document so that it does not block editor input latency. The only
+		// downside to this is that changing a character adjacent to a bracket may appear white for
+		// a brief period.
+		this.parseQueue.push(edits);
+		this.parseTimer.cancelAndSet(() => this.doParse(), 0);
```
Comment (alexdima) [inline src/vs/editor/common/model/bracketPairsTextModelPart/bracketPairsTree/bracketPairsTree.ts:117]: We should remove this and call `doParse()` before all public methods of the class.

**Review by alexdima — DISMISSED** [review]

**Comment by Tyriar** [comment]
@hediet ready for a review, this has the changes @alexdima and I discussed which should ensure the state is correct when public APIs are called by flushing the queue at the last moment.

**Comment by Tyriar** [comment]
Adding @rebornix so I can get this in before endgame

**Comment by alexdima** [comment]
@Tyriar I would feel most comfortable if we can also get @hediet's review here, as he authored/owns that code.

**Review by rebornix — APPROVED** [review]

**Comment by Tyriar** [comment]
I pinged @hediet about it to make sure he checks it out

**Comment by hediet** [comment]
The incremental bracket pair parser expects that the given edits precisely describe the changes between the state of the text model it saw the last time and the current state of the text model.
Queuing the changes and then processing them in the order they have been queued breaks this assumption, as the parser only sees the final state of the text model when processing these intermediate changes.

Imagine you start with this document:
```

}
```

Now the user
* (1) adds `{` at line 1, column 1 (new length of the replaced text: 1 column)
* (2) adds `\n` at line 1, column 1 (new length of the replaced text: 0 columns, 1 line)

The state after (1) is:
```
{
}
```

The state after (2) is:
```

{
}
```

However, when the parser processes (1), it already sees the state (2).
The parser tries to reparse the text node at that position (line 1, column 1 to line 1, column 2) and still finds an empty line - thus the bracket tree is not touched.
Then it processes the `\n` change and again does not touch the brackets 🐛

**Pull request merged**